Fix issue #177

tai271828 opened the pull request:
* Please merge https://github.com/solvcon/solvcon/pull/179 first to make the travis CI build work again.
* The issue #177 has described the rationale. Please refer to the inline comments for details.

## Steps to Verify

1. Run the driving script

``
$ cd <SOLVCON/sandbox/gas/tube/go
$ ./go run
``

2. We will get the expected result 1

3. Tweak the driving script to use an undefined boundary condition type. For example, replace ``b'wall'`` by ``somebc``. Run the driving script again.

4. We will get the expected result 2


## Expected Result
1. The driving script will run through the CESE march, and generate the desired data.
2. The driving script will crash because

``
AttributeError: ('[BC#0 "b\'wall\'": 1704 faces with 0 values]', 'soln', "'BC' object has no attribute 'soln'")
``

and you should get some prompt like:

```
*** *** 
*** ***   One of the boundary conditions is generic boundary type.
*** ***   Unexpected boundary conditions may be assigned.
*** *** 
*** ***   Suggestion: Check your boundary condition mapper of your driving script.
*** *** 
```

tai271828 commented on sandbox/gas/tube/go:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env python2.7
+#!/usr/bin/env python3.5
```
Bump the version because SOLVCON has migrated to python 3. (and 3.5 specifically so far for [the python11 compatibility issue](https://github.com/solvcon/solvcon/commit/2329189d383988c5ea0cb5705cfb69fc5e4c4eda))

yungyuc replied:
SOLVCON should eventually be compatible to Python 3.6.  We should be generic for Python 3 in driving script.

tai271828 replied:
Fixed. Thanks for the comment.

tai271828 commented on sandbox/gas/tube/go:
```diff
@@ -99,7 +99,7 @@ class SodtubeMesher(object):
     Circle(3) = {4,1,5};
     Circle(4) = {5,1,2};
-    // connect the arc to get a circle  
+    // connect the arc to get a circle
```
pep8: removing trailing space.

tai271828 commented on sandbox/gas/tube/go:
```diff
@@ -134,9 +134,9 @@ class SodtubeMesher(object):
-        'right': (sc.bctregy.GasNonrefl, {},),
+        b'wall': (sc.bctregy.GasWall, {},),
+        b'left': (sc.bctregy.GasNonrefl, {},),
+        b'right': (sc.bctregy.GasNonrefl, {},),
```
python2 to 3 compatibility issue. Make sure bytes string be used to match the desired elements of boundary condition mapper.

tai271828 commented on sandbox/gas/tube/go:
```diff
@@ -195,7 +195,7 @@ def tube_base(casename=None,
     cse.defer(gas.ProgressHook, linewidth=ssteps / psteps, psteps=psteps)
     cse.defer(gas.CflHook, fullstop=False, cflmax=10.0, psteps=ssteps)
-    cse.defer(gas.MeshInfoHook, psteps=ssteps)
+    cse.defer(gas.MeshInfoHook, psteps=ssteps, show_bclist=True)
```
Only for verbose information to catch and verify the fix. It is not related to fix the compatibility directly.

tai271828 commented on solvcon/parcel/gas/inout.py:
```diff
@@ -65,12 +65,28 @@ def __init__(self, cse, show_bclist=False, perffn=None, **kw):
+            self.info("  Suggestion: Check your boundary condition mapper of "
+                      "your driving script.\n")
+            self.info("\n")
+
```
By default, a generic class BC (boundary condition abstract class) will be used if there is no expected boundary type was found in the boundary condition mapper. However, the driving script of gas crashed because BC class has no attribute like soln:

```
AttributeError: ('[BC#0 "b\'wall\'": 1704 faces with 0 values]', 'soln', "'BC' object has no attribute 'soln'")
```

Type check so we could catch similar issue in the future. If all of the boundary condition type could be found from the bcoundary condition mapper, the driving script will show nothing. If there is any type is the abstract BC type, the driving script will throw out the prompt. It looks like:

```
......
*** ****************************************************************************
*** *** Start run_preloop ... 
*** *** ************************************************************************
*** *** Steps 0/600
*** *** Block information:
*** ***   [Block (3D/centroid): 1160 nodes, 8934 faces (1832 BC), 4009 cells]
*** *** 
*** ***   One of the boundary conditions is generic boundary type.
*** ***   Unexpected boundary conditions may be assigned.
*** *** 
*** ***   Suggestion: Check your boundary condition mapper of your driving script.
*** *** 
*** *** Traceback (most recent call last):
  File "./go", line 234, in <module>
    sc.go()
  File "/home/tai271828/my-local-installation/miniconda3/solvcon-dev/lib/python3.5/site-packages/solvcon/cmdutil.py", line 190, in go
    cmd()
  File "/home/tai271828/my-local-installation/miniconda3/solvcon-dev/lib/python3.5/site-packages/solvcon/command.py", line 1094, in __call__
    func(submit=False, **funckw)
  File "/home/tai271828/my-local-installation/miniconda3/solvcon-dev/lib/python3.5/site-packages/solvcon/case.py", line 1085, in simu
    cse.run(level=runlevel)
  File "/home/tai271828/my-local-installation/miniconda3/solvcon-dev/lib/python3.5/site-packages/solvcon/case.py", line 841, in run
    self._run_preloop()
  File "/home/tai271828/my-local-installation/miniconda3/solvcon-dev/lib/python3.5/site-packages/solvcon/case.py", line 876, in _run_preloop
    self.solver.solverobj.apply_bc()
  File "/home/tai271828/my-local-installation/miniconda3/solvcon-dev/lib/python3.5/site-packages/solvcon/parcel/gas/solver.py", line 149, in apply_bc
    self.call_non_interface_bc('soln')
  File "/home/tai271828/my-local-installation/miniconda3/solvcon-dev/lib/python3.5/site-packages/solvcon/solver.py", line 424, in call_non_interface_bc
    getattr(bc, name)(*args, **kw)
AttributeError: ('[BC#0 "b\'wall\'": 1704 faces with 0 values]', 'soln', "'BC' object has no attribute 'soln'")
```

tai271828 commented:
The Travis CI build failed because the issue #178. I have merged the fix of #178 and build with my branch. Please refer to the successful build [here](https://travis-ci.org/tai271828/solvcon/builds/189811054).

yungyuc commented on solvcon/parcel/gas/inout.py:
```diff
@@ -65,12 +65,28 @@ def __init__(self, cse, show_bclist=False, perffn=None, **kw):
         super(MeshInfoHook, self).__init__(cse, **kw)
 
     def preloop(self):
+        from ...boundcond import BC
```
`BC` is available in the `solvcon` top-level namespace.  This line is not necessary.

tai271828 replied:
Fixed. Thanks for this useful comment.

yungyuc commented on solvcon/parcel/gas/inout.py:
```diff
@@ -65,12 +65,28 @@ def __init__(self, cse, show_bclist=False, perffn=None, **kw):
 
+        # explicit info to prompt users to watch out issues like issue #177
+        for bc in blk.bclist:
+            if bc.__class__ is BC:
```
`solvcon` has be imported as `sc`.  `sc.BC` is a valid expression.

tai271828 commented:
Fixed the driving script and the sc.BC namespace according to the comment, and re-submitted. I also rebased the commits to have a better look of commit logs because PR https://github.com/solvcon/solvcon/pull/179 has been merged.

yungyuc commented on solvcon/parcel/gas/inout.py:
```diff
@@ -65,12 +65,27 @@ def __init__(self, cse, show_bclist=False, perffn=None, **kw):
+        for bc in blk.bclist:
+            if bc.__class__ is sc.BC:
+                flag_generic_bc = True
+        if flag_generic_bc:
```
To make it clearer, combine the 5 lines above (and line 68) into a one-liner:

```python
if any(bc.__class__ is sc.BC for bc in blk.bclist):
```

tai271828 replied:
Thanks for the comment. Using the built-in function is indeed a better way. Fixed.

yungyuc commented on solvcon/parcel/gas/inout.py:
```diff
@@ -65,12 +65,27 @@ def __init__(self, cse, show_bclist=False, perffn=None, **kw):
+                flag_generic_bc = True
+        if flag_generic_bc:
+            self.info("\n")
+            self.info("  One of the boundary conditions is generic boundary "
```
We only need to keep this line of message.  The following lines may be removed.

tai271828 replied:
Fixed.

yungyuc commented on solvcon/parcel/gas/inout.py:
```diff
@@ -65,12 +65,27 @@ def __init__(self, cse, show_bclist=False, perffn=None, **kw):
+            if bc.__class__ is sc.BC:
+                flag_generic_bc = True
+        if flag_generic_bc:
+            self.info("\n")
```
Please remove this new line.

tai271828 replied:
Fixed.

tai271828 commented:
Fixed the changes requested and update the PR again by merging the fixup. Please review this PR again. Thanks.

yungyuc commented on solvcon/parcel/gas/inout.py:
```diff
@@ -65,12 +65,18 @@ def __init__(self, cse, show_bclist=False, perffn=None, **kw):
         super(MeshInfoHook, self).__init__(cse, **kw)
 
     def preloop(self):
+        flag_generic_bc = False
```
This is redundant.

tai271828 replied:
Fixed. Thanks for the reminder. I scanned ``inout.py`` this time by the code checker ``flake8``. I fixed all coding style issues related to this PR. The other pep8 issues prompted by ``flake8`` are not fixed to make the code review be easier. They can be fixed later in another PR.

yungyuc replied:
Thanks.  It's a good idea to use a tool for checking the formatting.  Perhaps we should add that to travis.  There are many such tools and I don't know if flake8 is the best one.  I know there are a lot of legacy code in SOLVCON that will not pass the checks.  We need a proper suppression file.

Each time you re-submit the PR, the my comments are gone in the newly updated changesets.  I am not sure why it happens.

tai271828 replied:
Do you mean that your comments were gone in the "Files changed" session? I think that is because I merged each fix-up into its related commit. "git rebase -i" feature was used and then I merged the fixup into the previous commit to make the commit history look better. This action changes the UUID of the commit, so Github regards it as a new commit, and only shows the "files changed" comments and code change for the latest commits.

From my perspective, it's ok to me because your comments are reserved in this conversation session. Just to click "show outdated" to expand the old comments.

yungyuc replied:
I see.

tai271828 commented:
Re-submitted and ready to review.
